src/main/java/service/impl/ShopServiceImpl.java

```
@@ -0,0 +1,32 @@
+package service.impl;
+
+import java.util.List;
+import model.FruitTransaction;
+import service.ShopService;
+import strategy.OperationHandler;
+import strategy.OperationStrategy;
+
+public class ShopServiceImpl implements ShopService {
+
+    private final OperationStrategy operationStrategy;
+
+    public ShopServiceImpl(OperationStrategy operationStrategy) {
+        this.operationStrategy = operationStrategy;
+    }
+
+    @Override
+    public void process(List<FruitTransaction> transactions) {
+        if (transactions == null) {
+            throw new IllegalArgumentException("Transactions cannot be null");
+        }
+
+        for (FruitTransaction transaction : transactions) {
+            OperationHandler handler = operationStrategy.getHandler(transaction.getOperation());
```

You call `operationStrategy.getHandler(transaction.getOperation())` but do not guard against `transaction.getOperation()` being null nor rely on the strategy to throw an informative error. As noted in prior guidance, the OperationStrategy should validate inputs and fail fast: "OperationStrategyImpl: validate constructor `handlers` (non-null + defensive copy) and in `getHandler` throw informative error when `operation` is null or handler is missing (don’t return null)." If your strategy can still return null, at minimum the thrown error here (`No handler found for operation: ...`) should include transaction context (fruit, quantity, and ideally the original CSV line/index) so users can locate the bad row.